### PR TITLE
Disallow container mutation during lazy configuration

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
Fixes https://github.com/gradle/gradle-native/issues/709

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=lacasseio%2Flazy%2Fdisallow-container-modification)
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy [hunk 1]
```diff
@@ -1431,4 +1420,339 @@ abstract class AbstractDomainObjectCollectionSpec<T> extends Specification {
         1 * action.execute(a)
         0 * action.execute(_)
     }
+
+    abstract static class AbstractConfigurationFactory<T> implements Factory<T> {
```
Comment (ghale) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy]: I would consider pulling these classes out into a fixture class or something.  It's a lot of additional noise in an already noisy test class.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy [hunk 2]
```diff
@@ -1431,4 +1420,339 @@ abstract class AbstractDomainObjectCollectionSpec<T> extends Specification {
...
+    }
+
+    class AddAllLaterClosureFactory<T> extends AbstractConfigurationClosureFactory {
+        AddAllLaterClosureFactory() {
+            containerAllowsExternalProviders()
+        }
+
+        @Override
+        void call() {
+            container.addAllLater(new CollectionProviderInternal<T, Collection<T>>() {
```
Comment (ghale) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/AbstractDomainObjectCollectionSpec.groovy]: Is this anonymous type different than the anonymous type in the previous class?  If not, then let's pull it out and reuse the definition - then if there are changes to the interfaces, it doesn't require multiple mechanical updates.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectListTest.groovy [hunk 3]
```diff
@@ -460,4 +462,135 @@ class DefaultNamedDomainObjectListTest extends AbstractNamedDomainObjectCollecti
         iter.next() == "c"
         !iter.hasNext()
     }
+
+    class InsertActionFactory extends AbstractDomainObjectCollectionSpec.AbstractConfigurationActionFactory {
```
Comment (ghale) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultNamedDomainObjectListTest.groovy]: Same here - maybe hide these extra types in a fixture.

File: subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java [hunk 4]
```diff
@@ -144,7 +145,20 @@ public boolean addAll(Collection<? extends T> c) {
     }
 
     @Override
-    public void addLater(final Provider<? extends T> provider) {
+    public void addLater(Provider<? extends T> provider) {
+        assertMutable("addLater(Provider)");
+        if (provider instanceof Named) {
+            final Named named = (Named) provider;
+            String name = named.getName();
+            if (index.get(name) != null || index.getPending(name) != null) {
```
Comment (ghale) [inline subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java]: If coverage for this case (registering a provider with the same name as an existing provider or realized element) is missing, we should add some in the unit tests.

Comment (lacasseio) [reply subprojects/core/src/main/java/org/gradle/api/internal/DefaultNamedDomainObjectCollection.java]: I added test coverage for this.

File: subprojects/core/src/main/java/org/gradle/api/internal/collections/IterationOrderRetainingSetElementSource.java [hunk 5]
```diff
@@ -46,6 +46,10 @@ public boolean isSatisfiedBy(ValuePointer<T> pointer) {
 
     @Override
     public boolean add(T element) {
+        if (isRealizing()) {
+            return true;
```
Comment (lacasseio) [inline subprojects/core/src/main/java/org/gradle/api/internal/collections/IterationOrderRetainingSetElementSource.java]: 🔍 This tries to fix the difference between `ElementSource` implementation where an element been resolved would kick an add inside the container which would succeed and duplicate the element. This should be reviewed deeply to ensure we aren't breaking something. Don't hesitate to comment on test coverage that we could add as well.

File: subprojects/core/src/main/java/org/gradle/api/internal/project/BuildOperationCrossProjectConfigurator.java [hunk 6]
```diff
@@ -30,12 +31,7 @@
 public class BuildOperationCrossProjectConfigurator implements CrossProjectConfigurator {
 
     private final BuildOperationExecutor buildOperationExecutor;
-    private final ThreadLocal<Boolean> allowExecution = new ThreadLocal<Boolean>() {
-        @Override
-        protected Boolean initialValue() {
-            return true;
-        }
-    };
+    private final MutationGuard mutationGuard = new DefaultMutationGuard();
```
Comment (lacasseio) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/BuildOperationCrossProjectConfigurator.java]: ❓ I'm wondering if we can move this `MutationGuard` instance inside `DefaultGradle` instance and just have all `Project` fetch the guard from there and below, we could use `MutationGuards.of(project)` to get the guard. This assumes a single `Gradle` instance is constructed as `CrossProjectConfigurer` is instantiated once per build session.

File: subprojects/core/src/main/java/org/gradle/api/internal/project/CrossProjectConfigurator.java [hunk 7]
```diff
@@ -18,8 +18,9 @@
 
 import org.gradle.api.Action;
 import org.gradle.api.Project;
+import org.gradle.api.internal.WithMutationGuard;
 
-public interface CrossProjectConfigurator {
+public interface CrossProjectConfigurator extends WithMutationGuard {
```
Comment (lacasseio) [inline subprojects/core/src/main/java/org/gradle/api/internal/project/CrossProjectConfigurator.java]: ❓  Now that we removed the mutation guard from this class, would it make sense to move the mutation guard somewhere else that is instantiated once per build session?

**Comment by lacasseio** [comment]
Abandoning this PR

**Pull request closed without merging**